### PR TITLE
159 - File picker

### PR DESCRIPTION
**Pull request by latin-panda**
# Description
In this PR:
- Removes the old image picker because it's deprecated. 
- Completely removes support for deprecated library `android.support.*` (`android.enableJetifier=true`) and fixes style of buttons that were depending on `Widget.AppCompat`.
- Adds new support for Webview's file picker. 
- Adds support for recording if selected instead of picking from file system

Ticket: https://github.com/medic/cht-android/issues/159

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: src/main/res/drawable/standard_button.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (latin-panda) [inline src/main/res/drawable/standard_button.xml:1]: Newer Android versions is shaping the buttons like "oval", we used to have it like square with slightly rounded corders. 
No way to shape buttons with rounded corners as we do in CSS. Here we create a drawable and use it as button's background, this is supported in Android 5+. 
So to keep consistency between Android versions and as it was originally.

**Comment by latin-panda** [comment]
@dianabarsan when you have time, can you please review? Thanks!

**Comment by latin-panda** [comment]
Today I had a chat with @garethbowen about this feature, for the user convenience we're going to also offer the option of recording. 
The file size limit will be configurable in  cht-core's app-settings per project (no per form). There are projects that recently upgraded the CHW's phones to higher end and they might want to support bigger files. 

@dianabarsan  please hold the review while I add the additional support, sorry for the back and forth :)

File: src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -63,6 +63,17 @@
       android:screenOrientation="portrait"
       android:configChanges="orientation|screenSize"
       android:launchMode="singleTop"/>
+
+    <provider
```
Comment (latin-panda) [inline src/main/AndroidManifest.xml:67]: Allows the app to create files in its directory.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		this.context.startActivityForResult(chooserIntent, RequestCode.FILE_PICKER_ACTIVITY.getCode());
+	}
+
+	private Intent getCaptureIntent(String mimeType) {
+		if (mimeType.contains("video/")) {
+			return new Intent(ACTION_VIDEO_CAPTURE);
+		}
+
+		if (mimeType.contains("audio/")) {
+			return new Intent(RECORD_SOUND_ACTION);
```
Comment (latin-panda) [inline src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:137]: This one is tricker to test, in my experience not all recording apps are set to open with this intent. 
I found that one called: "Audio Recorder" by "gitlab.com/axet" works with this.

**Comment by latin-panda** [comment]
Hi @dianabarsan, when you have time, can you please have a look? 
I've tested in Android 9 and 12 with real devices. And tested Android 7 in emulator.

File: src/main/AndroidManifest.xml (hunk 2 above)
Comment (latin-panda) [reply src/main/AndroidManifest.xml:67]: To take a picture, we need to create a file to store the pic. 
To record video, we don't need to create a file 🤷‍♀️

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 3 above)
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:137]: I got this app from playstore.

**Comment by dianabarsan** [comment]
@latin-panda I'm sorry I didn't get to this today :( I'm scheduling it for Monday. Sorry for the wait.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		}
+
+		trace(this, "FilePickerHandler :: Starting activity with intents: chooser=%s, picker=%s, capture=%s",
+			chooserIntent, pickerIntent, captureIntent);
+
+		this.context.startActivityForResult(chooserIntent, RequestCode.FILE_PICKER_ACTIVITY.getCode());
+	}
+
+	private Intent getCaptureIntent(String mimeType) {
+		if (mimeType.contains("video/")) {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:132]: I added two issues in cht-core related to this: 
https://github.com/medic/cht-core/issues/7531
https://github.com/medic/cht-core/issues/7530

So the only problem remains the invalid preview. 
Without the form actually saving something and without the preview, it's hard to check that this feature works. 

I was uploading / recording an `mp4`.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		pickerIntent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
+		Intent captureIntent = null;
+
+		if (mimeTypes.length == 1) {
+			String mimeType = mimeTypes[0];
+			pickerIntent.setType(mimeType);
+			captureIntent = getCaptureIntent(mimeType);
+		} else {
+			pickerIntent.setType("*/*");
+			pickerIntent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
```
Comment (dianabarsan) [inline src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:116]: I'm not sure how this should work. I created a form with a binary input and an empty media type, and I got: 
Screenshot_20220221-102944

File: src/main/res/drawable/standard_button.xml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+<shape xmlns:android="http://schemas.android.com/apk/res/android" android:shape="rectangle">
+  <solid android:color="#D6D7D7" />
```
Comment (dianabarsan) [inline src/main/res/drawable/standard_button.xml:3]: Is this related to this feature? 
Also, the button ends up looking like this on Android 6: 
image

White text on light grey is hard to see.

File: src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+	@Test
+	public void processResult_withNullUri_finishesPicker() {
+		try (MockedStatic<MedicLog> medicLogMock = mockStatic(MedicLog.class)) {
+			//> GIVEN
+			ValueCallback<Uri[]> filePickerCallbackMock = mock(ValueCallback.class);
+			FilePickerHandler filePickerHandler = new FilePickerHandler(contextMock);
+			filePickerHandler.setFilePickerCallback(filePickerCallbackMock);
+
+			Intent intent = mock(Intent.class);
+			when(intent.getData()).thenReturn(null);
```
Comment (dianabarsan) [inline src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java:120]: I'm not sure what should happen in webapp at this point. If I cancel the upload, I get an invalid image now.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice improvement! 
I'm fairly useless when reviewing Java code, but I did give this a go and left some feedback inline.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 5 above)
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:116]: Can you please the xml for this input? Thanks!

File: src/main/res/drawable/standard_button.xml (hunk 6 above)
Comment (latin-panda) [reply src/main/res/drawable/standard_button.xml:3]: I'll fix this button. Is the screen background dark gray because of the phone theme? Was it like that before? I wonder if that's a regression as well.

File: src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java (hunk 7 above)
Comment (latin-panda) [reply src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java:120]: 🤔 I wonder if webapp can detect an invalid file and ignore it. I'll have a look

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:132]: Thanks for the ticket! Yes, the edit fails.
In theory it should still save the `mp4` in the form doc's attachment in CouchDB. I'll have a look again

File: src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java (hunk 7 above)
Comment (latin-panda) [reply src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java:120]: Like when bytes are zero

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 5 above)
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:116]:
```
<bind nodeset="/photos-on-android/my-any-file" type="binary"/>
....
<upload ref="/photos-on-android/my-any-file" mediatype="">
			<hint>Select a whatever?</hint>
			<label>File widget</label>
		</upload>
```

File: src/main/res/drawable/standard_button.xml (hunk 6 above)
Comment (dianabarsan) [reply src/main/res/drawable/standard_button.xml:3]: The background was always like this, as far as I remember.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	 */
+	private void sendDataToWebapp(Intent intent) {
+		if (this.filePickerCallback == null) {
+			warn(this, "FilePickerHandler :: Cannot send data back to webapp, filePickerCallback is null.");
+			return;
+		}
+
+		Uri uri = getSelectedFileUri(intent);
+		trace(this, "FilePickerHandler :: Sending data back to webapp, URI: %s", uri);
+		this.filePickerCallback.onReceiveValue(uri == null ? null : new Uri[] { uri });
```
Comment (latin-panda) [inline src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:80]: To add more info: I just double checked about sending `null` in `onReceiveValue`, this is the [recommended way](https://developer.android.com/reference/android/webkit/WebChromeClient#onShowFileChooser(android.webkit.WebView,%20android.webkit.ValueCallback%3Candroid.net.Uri[]%3E,%20android.webkit.WebChromeClient.FileChooserParams)) in the android site.

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+				warn(this, "FilePickerHandler :: Cannot create temporary file for image capture: %s", ex);
+			}
+		}
+
+		trace(this, "FilePickerHandler :: No capture intent for MIME type: %s", mimeType);
+		return null;
+	}
+
+	private void createTempImageFile() throws Exception {
+		String imageFileName = "medic-" + System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:162]: Done

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 5 above)
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:121]: Done

File: src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java (hunk 7 above)
Comment (latin-panda) [reply src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java:120]: I've mitigated this one in the android app, but still needs the fix in webapp.

File: src/main/res/drawable/standard_button.xml (hunk 6 above)
Comment (latin-panda) [reply src/main/res/drawable/standard_button.xml:3]: I've fixed the button

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:137]: I can save the attachments: 
Screen Shot 2022-02-23 at 4 57 50 pm

And click on any of these to see the content, this is a video:
Screen Shot 2022-02-23 at 4 58 09 pm

But yes, it needs the fixes in webapp to properly edit the form and preview the files

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 5 above)
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:121]: The behaviour is: if there isn't a mime type, it shouldn't let to select a file.

**Comment by latin-panda** [comment]
Hi @dianabarsan, can you please have a look again? Thank you

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java (hunk 5 above)
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:121]: Ok, so right now, if I click on this field, nothing should happen, right?

File: src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+			}
+			setTempFile(null);
+		}
+
+		return uri;
+	}
+
+	private void startFileCaptureActivity(FileChooserParams fileChooserParams) {
+		String[] mimeTypes = cleanMimeTypes(fileChooserParams);
+		if (mimeTypes.length == 0) {
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:102]: 🤔🤔🤔🤔🤔🤔 weird, I tested with that input, but didn't got this. 
I'm assuming this is Android 6, right ?

[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:102]: Yes, Android 6 with latest Chrome.

[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:102]: I'm looking at it again, sorry about that

[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java:102]: This is what I get on Android 12: 
image

[the remark this answers is not in the document]
Comment (latin-panda) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java]: Sorry for the back and forth, I didn't understand before.
I've pushed the code to allow opening file picker when no mime type is defined

[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/main/java/org/medicmobile/webapp/mobile/FilePickerHandler.java]: I think, with the recent changes, opening just triggers the "Files" app intent, while in the browser, I get the "take picture", "record video" and "Files" intents (as displayed in the screenshot above). 
Leaving this up to quality assurance.

**Review by dianabarsan — APPROVED** [review]
Great work! Thanks a lot!
